controller/src/main/java/org/jboss/as/controller/operations/global/EnhancedSyntaxSupport.java

```diff
@@ -0,0 +1,146 @@
+/*
+ * JBoss, Home of Professional Open Source
+ * Copyright 2015, Red Hat, Inc., and individual contributors as indicated
+ * by the @authors tag.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.jboss.as.controller.operations.global;
+
+import static org.jboss.as.controller.descriptions.ModelDescriptionConstants.ATTRIBUTES;
+import static org.jboss.as.controller.descriptions.ModelDescriptionConstants.DEFAULT;
+
+import java.util.Locale;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.jboss.as.controller.OperationContext;
+import org.jboss.as.controller.OperationFailedException;
+import org.jboss.as.controller.PathAddress;
+import org.jboss.as.controller.descriptions.DescriptionProvider;
+import org.jboss.as.controller.logging.ControllerLogger;
+import org.jboss.as.controller.registry.ImmutableManagementResourceRegistration;
+import org.jboss.as.controller.registry.Resource;
+import org.jboss.dmr.ModelNode;
+import org.jboss.dmr.ModelType;
+
+/**
+ * @author Tomaz Cerar (c) 2015 Red Hat Inc.
+ */
+class EnhancedSyntaxSupport {
+
+    static final Pattern ENHANCED_SYNTAX_PATTERN = Pattern.compile("(.*[\\.\\[\\]].*)");
+    static final Pattern BRACKETS_PATTERN = Pattern.compile("(.*)\\[(-?\\d+)\\]");
+    static final Pattern EXTRACT_NAME_PATTERN = Pattern.compile("^(.*?)[\\.\\[].*");
+
+    static void resolveAttribute(OperationContext context, ModelNode operation, String attributeName, boolean defaults, ImmutableManagementResourceRegistration registry, boolean enhancedSyntax) throws OperationFailedException {
+        final Resource resource = context.readResource(PathAddress.EMPTY_ADDRESS, false);
+        final ModelNode subModel = resource.getModel();
+        if (enhancedSyntax){
+            context.getResult().set(resolveEnhancedSyntax(attributeName,subModel));
+        }else if (subModel.hasDefined(attributeName)) {
```

Comment:
There's a problem here re the " || !defaults" check from the old L194 of ReadAttributeHandler. That check is not present in the old L170 of ReadAttributeHandler, so there's a difference in the handling between those two cases. The difference is valid, as in the L170 case, there is no AttributeAccess, so if we need to deal with the fact that the attribute name might be bogus. If we added the " || !defaults" check to L170, then when defaults == false, we'd just return undefined instead of dropping into the code that detects the fact that the attribute name is bogus.

My proposal:

1) We change your current L171 of ReadAttributeHandler to no longer call this method, and just do

```
throw new OperationFailedException(ControllerLogger.ROOT_LOGGER.unknownAttribute(attributeName));
```

Basically, we no longer support attributes not registered in the ManagementResourceRegistration.

2) Then L52 ^^^ has the " || !defaults" restored.